src/pudl/extract/extractor.py

```diff
@@ -11,18 +13,239 @@
     op,
 )
 
+import pudl
+
+StrInt = str | int
+PartitionSelection = list[StrInt] | tuple[StrInt] | StrInt
+
+logger = pudl.logging_helpers.get_logger(__name__)
+
+
+class GenericMetadata:
+    """Load generic metadata from Python package data.
+
+    When metadata object is instantiated, it is given ${dataset} name and it
+    will attempt to load csv files from pudl.package_data.${dataset} package.
+
+    It expects the following kinds of files:
+    * column_map/${page}.csv currently informs us how to translate input column
+      names to standardized pudl names for given (partition, input_col_name).
+      Relevant page is encoded in the filename.
+    """
+
+    def __init__(self, dataset_name: str):
+        """Create Metadata object and load metadata from python package.
+
+        Args:
+            dataset_name: Name of the package/dataset to load the metadata from.
+            Files will be loaded from pudl.package_data.${dataset_name}
+        """
+        self._dataset_name = dataset_name
+        self._pkg = f"pudl.package_data.{dataset_name}"
+        column_map_pkg = self._pkg + ".column_maps"
+        self._column_map = {}
+        for res_path in importlib.resources.files(column_map_pkg).iterdir():
+            # res_path is expected to end with ${page}.csv
+            if res_path.suffix == ".csv":
+                column_map = self._load_csv(column_map_pkg, res_path.name)
+                self._column_map[res_path.stem] = column_map
+
+    def get_dataset_name(self) -> str:
+        """Returns the name of the dataset described by this metadata."""
+        return self._dataset_name
+
+    @staticmethod
+    def _load_csv(package: str, filename: str) -> pd.DataFrame:
+        """Load metadata from a filename that is found in a package."""
+        return pd.read_csv(
+            importlib.resources.files(package) / filename, index_col=0, comment="#"
+        )
+
+    @staticmethod
+    def _get_partition_selection(partition: dict[str, PartitionSelection]) -> str:
+        """Grab the partition key."""
+        partition_names = list(partition.keys())
+        if len(partition_names) != 1:
+            raise AssertionError(
+                f"Expecting exactly one attribute to define this partition (found: {partition})"
+            )
+
+        partition_name = partition_names[0]
```

Comment:
A file might currently be defined by more than one partition (e.g., form _and_ year) - PHMSA is a definite edge case, if it's helpful to look at how the partitions are structured there: https://zenodo.org/records/10231982

Reply:
Ignore me, I see you've handled this below!